File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> anagramGroup;
        int n = strs.size();
        for(int i=0;i<n;i++){
            vector<int> charCount(26, 0);
            for(char c: strs[i]){
                charCount[c - 'a']++;
            }
            string key;
            for(int charVal: charCount){
                key = key + to_string(charVal)+"#";
            }
            anagramGroup[key].push_back(strs[i]);
        }
        vector<vector<string>> result;
        for(auto x: anagramGroup){
            result.push_back(x.second);
        }
        return result;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        int n = strs.length;
        Map<String, List<String>> groupAnagrams = new HashMap<>();
        for(String word: strs){
            int charCount[] = new int[26];
            for(char c: word.toCharArray()){
                charCount[c - 'a']++;
            }
            StringBuilder sb = new StringBuilder();
            for(int charVal: charCount){
                sb.append(charVal);
                sb.append("#");
            }
            String key = sb.toString();
            if(!groupAnagrams.containsKey(key)){
                groupAnagrams.put(key, new ArrayList<>());
            }
            groupAnagrams.get(key).add(word);
        }
        List<List<String>> result = new ArrayList<>();
        for(Map.Entry<String, List<String>> entry: groupAnagrams.entrySet()){
            result.add(entry.getValue());
        }
        return result;
    }
}